cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -1016,6 +1055,24 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
 	return nil
 }
 
+// checkDataAndSavaPairing make sure that ubuntu-data and ubuntu-save
+// come from the same install by comparing secret markers in them
+func checkDataAndSavaPairing(rootdir string) (bool, error) {
+	// read the secret marker file from ubuntu-data
+	markerFile1 := filepath.Join(dirs.SnapFDEDirUnder(rootdir), "marker")
+	marker1, err := ioutil.ReadFile(markerFile1)
+	if err != nil {
+		return false, err
+	}
+	// read the secret marker file from ubuntu-save
+	markerFile2 := filepath.Join(dirs.SnapFDEDirUnder(boot.InitramfsUbuntuSaveDir), "marker")
+	marker2, err := ioutil.ReadFile(markerFile2)
+	if err != nil {
+		return false, err
+	}
+	return subtle.ConstantTimeCompare(marker1, marker2) == 1, nil
```

Comment:
why do we use subtle here? are we concerned about hardware observation of what the secrets are due to how long it takes to measure their equality?

Reply:
I'm not super concerned about an attack, most likely you would have to have an unencrypted data but open an encrypted save which I think is not something that state machine lets one do. OTOH is a good habit to use subtle when comparing secrets.